### ISSUE TITLE
[PATCH] basic authentication support for twisted.web

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10448]
|<img alt="Yangs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Yangs reported|
|-|-|
|Trac ID|trac#696|
|Type|enhancement|
|Created|2004-08-30 05:06:02Z|
|Branch|https://github.com/twisted/twisted/tree/basicauth-web-696-3|

Attachments:

* [basicguard.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4c8/4c87e5e9fc1a564781c398618687a2f5cad0abb8/3da64c37905d4ebfa6f5aa779b656e26edf06e6f.py) (3059 bytes) - added by Yangs on 2004-08-30 05:06:01Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__696 696
type__enhancement enhancement
reporter__Yangs Yangs
priority__normal normal
milestone__ 
branch__branches_basicauth_web_696_3 branches/basicauth-web-696-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1093842362000000 1093842362000000
changetime__1213792164000000 1213792164000000
version__ 
owner__ 
cc__jknight cc__Yangs cc__dreid
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Yangs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Yangs commented|
|-|-|

```
#!html
<pre>
Hi.

I've implemented lightweight session-less basic authentication support for
twisted.web.

Attached.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Assigning to jknight so he can work out how this fits in the rewrite.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="chewy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|chewy commented|
|-|-|

(In [22889]) Branching to 'basicauth-web-696'

**Comment by twisted-trac** [comment]
|<img alt="chewy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|chewy set status to `closed`|
|-|-|

so, after a long discussion (see ticket:2042), went ahead and pulled web2.auth into web.auth so both web and web2 have http header-based authentication . 

docstrings and example code in doc/web/examples/webguard.py 

also of note, added an addHeader stub on http.request (pending completion of #7478)

**Comment by twisted-trac** [comment]
|<img alt="chewy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|chewy set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="chewy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|chewy removed owner<br>chewy set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_7"></a>|@jyknight commented|
|-|-|

It's harder to review this change than it ought to be, since the file copies weren't done with svn cp, and thus don't have copyfrom information, and thus svn diff doesn't show the changes.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to chewy|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_9"></a>|@dreid commented|
|-|-|

Please address the file copy issue jknight brought up.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [23769]) Branching to 'basicauth-web-696-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Um.  Actually, svn diff just doesn't follow moves.  The auth package was moved with `svn mv´ and do have copy-from information.  It's just really hard to figure out what changed with svn when there's a rename.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [23828]) Branching to 'basicauth-web-696-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [23874]) Branching to 'basicauth-web-696-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I think this is mostly done.  I got rid of the `twisted.web2` changes since the new code in `twisted.web` isn't completely compatible with what was there and I don't want to go to the effort to make it compatible.  I also made the whole auth package private and only exposed the tasty parts.  I also deleted a bunch of deprecated code from `twisted.web.guard`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

 * The example doesn't work for me. Here is what I've got when I run it: 
```
Traceback (most recent call last):
  File "/home/therve/Projets/Twisted/trunk/twisted/protocols/basic.py", line 231, in dataReceived
    why = self.lineReceived(line)
  File "/home/therve/Projets/Twisted/trunk/twisted/web/http.py", line 1197, in lineReceived
    self.allContentReceived()
  File "/home/therve/Projets/Twisted/trunk/twisted/web/http.py", line 1253, in allContentReceived
    req.requestReceived(command, path, version)
  File "/home/therve/Projets/Twisted/trunk/twisted/web/http.py", line 687, in requestReceived
    self.process()
--- <exception caught here> ---
  File "/home/therve/Projets/Twisted/trunk/twisted/web/server.py", line 150, in process
    self.render(resrc)
  File "/home/therve/Projets/Twisted/trunk/twisted/web/server.py", line 157, in render
    body = resrc.render(self)
  File "/home/therve/Projets/Twisted/trunk/twisted/web/_auth/wrapper.py", line 60, in render
    challenge = fact.getChallenge(request.client)
  File "/home/therve/Projets/Twisted/trunk/twisted/web/_auth/digest.py", line 329, in getChallenge
    o = self.generateOpaque(c, request.getClientIP())
```
 * it would be nice if the example called startLogging too
 * in test_httpauth, the challenge variables defined at module level doesn't seem to be used (except cnonce, but there is a class level variable with the same effect). They should probably be deleted.

More after lunch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @exarkun|
|-|-|

digest.py:
 * md5-sess is not tested in calcHA1
 * auth-int is not tested in calcResponse
 * `if pszNonceCount and pszCNonce` is not tested in calcResponse
 * the implements of DigestedCredentials are not tested
 * the lower() calls in DigestedCredentials.checkPassword and checkHash are not tested
 * the implements of DigestCredentialFactory is not tested
wrapper.py
 * the error in generateWWWAuthenticate is not tested
 * in HTTPAuthSessionWrapper.getChildWithDefault, `if factory is None` is not tested
basic.py:
 * the first except in decode is a bit broad, and not tested

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [23880]) explicit tests for calcHA1, including all three of md5, md5-sess, and sha

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [23881]) factor H(A2) calculation into separate function; explicit tests for it, including auth-int and auth QoP settings

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [23882]) explicit tests for calcResponse, including the conditional nc/cnonce/qop hashing

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [23883]) move IUsernameDigestHash to a public location; verify that it is provided by DigestedCredentials

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [23884]) fix the rest of digest.py

  * test that `BasicCredentialFactory` and `DigestCredentialFactory` implement `ICredentialFactory`
  * test that algorithms are considered after lowercasing
  * test that md5 is the default algorithm

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [23885]) get rid of the big try/except in generateWWWAuthenticate - it's pointless
refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [23886]) test `if factory is None:` case

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [23887]) fix the test for `+ '==='`; add a test for decoding failing anyway.  base64 in python is way too hard.

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun removed owner|
|-|-|

Everything mentioned in the review should be fixed now.  I noticed that `md5-sess` wasn't really fully implemented by `DigestCredentialFactory` so I changed its docs to point out that it should not be used.  It was fully implemented by all the hash functions though, so I left support for it there and added tests.  Similarly, `auth-int` isn't really supported by `DigestCredentialFactory`, but it always puts `auth` in its challenge, so this shouldn't break anything.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

I missed the first part of the review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [23889]) integrationy test between HTTPAuthSessionWrapper and ICredentialFactory, and a fix to make it pass.  makes this actually work.

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

(In [23890]) get rid of old globals

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Okay, **now** I think everything is addressed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun commented|
|-|-|

This was a duplicate of [#7077](https://github.com/twisted/twisted/issues/7077), it seems.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_31"></a>|@therve set owner to @exarkun|
|-|-|

 * pyflakes:
```
twisted/web/_auth/digest.py:15: 'Interface' imported but unused
```
 * I feel that addHeader should be a method of the `Header` class instead of putting the logic in the request

digest.py
 * the `strip('\n')` in generateOpaque is not tested
 * I think that twisted.python.randbytes would be the good thing to use instead of the random module
 * in decode, if unq returns `s[1:-1]` nothing fails

wrapper.py:
 * in render/quoteString, the second replace is not tested
 * the `return 'Unauthorized' is not tested
 * HTTPAuthSessionWrapper._loginSucceeded isn't documented, _loginFailed docstring is not complete and not well not formated, _selectParseHeader is not well formated.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

(In [23916]) address review feedback

refs [#10448](https://github.com/twisted/twisted/issues/10448)

  * fix pyflakes warning
  * remove `Request.addHeader` and add `Header.addRawHeader`; adjust DummyRequest to suit
  * more tests for `generateOpaque`
  * switch from `random` to `twisted.python.randbytes`
  * more tests for `DigestCredentialFactory.decode` quote handling
  * more tests for `UnauthorizedResource.render`
  * maintainer docs for `HTTPAuthSessionWrapper` private callback methods

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun removed owner|
|-|-|

Thanks for the review.  All points addressed, I believe.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_34"></a>|@therve set owner to @exarkun|
|-|-|

```
twisted/web/_auth/digest.py:13: 'sys' imported but unused
```

 * docstring of wrapper module is invalid

Please merge once this fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

(In [23948]) remove unused import

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

(In [23949]) fix markup in module docstring

refs [#10448](https://github.com/twisted/twisted/issues/10448)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun set status to `closed`|
|-|-|

(In [23950]) Merge basicauth-web-696-3

Author: dreid, exarkun
Reviewer: therve
Fixes: [#10448](https://github.com/twisted/twisted/issues/10448)

Add HTTP Basic and Digest authentication support for twisted.web
in the form of a guard-style resource wrapper with a twisted.cred
backend.

`DummyRequest` has been expanded slightly to more closely resemble
the real request class.  The long-since deprecated guard code in
twisted.web.guard has been removed and replaced with the new guard
API.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
